spec/controllers/offers_controller_spec.rb

```diff
@@ -1,64 +1,272 @@
 require 'rails_helper'
 
 RSpec.describe OffersController, type: :controller do
+  render_views
+
+  shared_examples 'it does not show global announcements' do
+    it 'does not show global announcements' do
+      expect(response).to_not render_template(:partial => '_global-announcements')
+    end
+  end
+
+  shared_examples 'an unsuccessful request' do
+    it 'is unsuccessful' do
+      expect(response).not_to be_successful
+    end 
+  end
+
+  shared_examples 'a successful request' do |text|
+    it 'is successful' do
+      expect(response).to be_successful
+    end
+
+    it 'renders the correct page' do
+      expect(response.body).to include(text)
+    end
+  end
+  
   describe 'GET #index' do
-    it 'works' do
-      pending 'TODO'
-      fail
+    let!(:offer) { create(:offer, name: 'Free books', user: user) }
+    let(:user) { create(:user) }
+    
+    before { get :index }
+    
+    it_behaves_like 'a successful request', 'Free books'
+
+    it 'assigns offers' do
+      expect(assigns(:offers)).to eq([offer])
     end
+
+    it_behaves_like 'it does not show global announcements'
   end
 
   describe 'GET #show' do
-    let!(:user) { create(:user) }
-    let!(:offer) { create(:offer, user: user) }
+    let(:user) { create(:user) }
+    let!(:offer) { create(:offer, name: 'Free coffee', user: user) }
 
-    it 'works with just a numeric id as param' do
-      get :show, params: { id: offer.id }
-      expect(response).to be_successful
-      expect(assigns(:offer)).to eq(offer)
+    before { get :show, params: { id: offer_id } }
+
+    context 'with a numeric id as param' do
+      let(:offer_id) { offer.id } 
+      
+      it_behaves_like 'a successful request', 'Free coffee'
+
+      it 'assigns the offer' do
+        expect(assigns(:offer)).to eq(offer)
+      end
+      
+      it_behaves_like 'it does not show global announcements'
     end
 
-    it 'works with a "5-my-offer-name" param too' do
-      expect(offer.to_param).to eq("#{offer.id}-#{offer.name.parameterize}")
-      get :show, params: { id: offer.to_param }
-      expect(response).to be_successful
-      expect(assigns(:offer)).to eq(offer)
+    context 'with a parameterized id' do 
```

Comment:
What is the utility of a parameterized id ie why can't we just use the offer id?

Reply:
it's so you can have prettier URLs. e.g. 
`https://helpwithcovid.com/offers/109-turtle-free-engineering-credit`
instead of
`https://helpwithcovid.com/offers/109`